### ISSUE TITLE
Nested derivate functions

### ISSUE THREAD

**Issue by ovainola** [JuliaDiff/ForwardDiff.jl#31]
Hi, I got my code working with the help of last issue: https://github.com/JuliaDiff/ForwardDiff.jl/issues/30. Though I found a another example how I would like to make my code to work

Like last example:
```julia
using ForwardDiff
f1(s) = (s/3)^2
f2(x) = x[1]^2 + exp(3*x[2]/sqrt(4))
function f(x)
    g = forwarddiff_gradient(f2, Float64, fadtype=:typed, n=2)
return f1(x[1]) + g(x)
end
g = forwarddiff_gradient(f, Float64, fadtype=:typed, n=2)
g([-1., 2.0])
```
ended up with error message:
```
LoadError: MethodError: `g` has no method matching g(::Array{ForwardDiff.GraDual{Float64,2},1})
while loading In[11], in expression starting on line 9

 in f at In[11]:6
 in g at C:\Users\ovax03\.julia\v0.4\ForwardDiff\src\typed_fad\GraDual.jl:188
```

I already found a workaround, but Im not sure should it work like this:
```julia
using ForwardDiff
f1(s) = (s/3)^2
f2(x) = x[1]^2 + exp(3*x[2]/sqrt(4))
function f(x)
    xd = map(t->t.v, x)
    g = forwarddiff_gradient(f2, Float64, fadtype=:typed, n=2)
    return f1(x[1]) + g(xd)
end
g = forwarddiff_gradient(f, Float64, fadtype=:typed, n=2)
g([-1., 2.0])
```
which gave me a hessia as expected:
```
2x2 Array{Float64,2}:
 -0.222222  0.0
 -0.222222  0.0
```

note: I'm working with finite element material models, which happen to have this kind of functions

**Comment by jrevels** [comment]
This is a really good use case, thanks for sharing. This is a problem with the new code in #27 as well - leaving a snippet here for later:

```julia
julia> using ForwardDiff

julia> f1(s) = (s/3)^2;

julia> f2(x) = x[1]^2 + exp(3*x[2]/sqrt(4));

julia> f(x) = f1(x[1]) + gradient(f2, x, Partials{2, Float64});

julia> g = gradient_func(f, Partials{2, Float64}, mutates=false);

julia> g([-1, 2.0])
ERROR: MethodError: `gradient` has no method matching gradient(::Function, ::Array{ForwardDiff.GradientNum{2,Float64,Tuple{Float64,Float64}},1}, ::Type{ForwardDiff.Partials{2,Float64}})
Closest candidates are:
  gradient{T<:Number}(::Function, ::Union{Array{T<:Number,1},T<:Number})
  gradient{T<:Number}(::Function, ::Union{Array{T<:Number,1},T<:Number}, ::Symbol)
  gradient{N,T}(::Function, ::Array{T,1}, ::Type{ForwardDiff.Partials{N,T}})
  ...
 in f at none:1
 in take_gradient! at /Users/jarrettrevels/.julia/ForwardDiff/src/fad_api.jl:85
 in gradf at /Users/jarrettrevels/.julia/ForwardDiff/src/fad_api.jl:110
```

**Comment by jrevels** [comment]
Rereading your code, I realize that `f` is a vector-valued function (`f`: R^2 --> R^2), meaning we want to take the *Jacobian*, not the gradient. You might try, then:

```julia
using ForwardDiff
f1(s) = (s/3)^2
f2(x) = x[1]^2 + exp(3*x[2]/sqrt(4))
function f(x)
    g = forwarddiff_gradient(f2, Float64, fadtype=:typed, n=2)
return f1(x[1]) + g(x)
end
j = forwarddiff_jacobian(f, Float64, fadtype=:typed, n=2)
j([-1., 2.0])
```

I'm not sure whether that will actually work or not - it seems the old code uses the same implementation to take the gradient as it does the Jacobian. The #27 implementation, however, handles this fine:

```julia
julia> using ForwardDiff

julia> f1(s) = (s/3)^2;

julia> f2(x) = x[1]^2 + exp(3*x[2]/sqrt(4));

julia> f(x) = f1(x[1]) + gradient(f2, x, Partials{2});

julia> j = jacobian_func(f, Partials{2}, 2, mutates=false);

julia> j([-1, 2.0])
2x2 Array{Float64,2}:
  1.77778    0.0
 -0.222222  45.1925
```

**Comment by ovainola** [comment]
Thanks for answering so soon! I couldn't the example code working but I'll get the idea you've implemented it in https://github.com/JuliaDiff/ForwardDiff.jl/pull/27 and if it's coming within couple of weeks I can live with that. My model converges quite nicely with newton at least with in 1D solution even if the jacobian might not be correctly evaluated.

**Comment by jrevels** [comment]
This now fixed on `master` as of the merging of #27.

**Comment by ovainola** [comment]
:+1:
